src/libraries/System.Net.Primitives/src/System/Net/IPNetwork.cs

```diff
@@ -0,0 +1,344 @@
+
+            if (IsAnyMaskBitOnForBaseAddress())
+            {
+                return _baseAddressHasSetBitsInMaskError;
+            }
+
+            return default;
+        }
+
+        /// <summary>
+        /// Method to determine whether any bit in <see cref="BaseAddress"/>'s variable/mask part is 1.
+        /// </summary>
+        private bool IsAnyMaskBitOnForBaseAddress()
+        {
+            // TODO: same as with Contains method - this can be made much easier and potentially more performant for IPv4 if IPAddress.PrivateAddress is made internal (currently private)
+            // to be discussed in the PR
+
+            Span<byte> addressBytes = stackalloc byte[GetAddressFamilyByteLength(BaseAddress.AddressFamily)];
+            if (!BaseAddress.TryWriteBytes(addressBytes, out int bytesWritten))
+            {
+                throw new UnreachableException();
+            }
+
+            var addressBitsCount = addressBytes.Length * bitsPerByte;
+
+            for (int addressBytesIndex = addressBytes.Length - 1, numberOfEndingBitsToBeOff = addressBitsCount - PrefixLength;
+                addressBytesIndex >= 0 && numberOfEndingBitsToBeOff > 0;
+                addressBytesIndex--, numberOfEndingBitsToBeOff -= bitsPerByte)
+            {
+                byte maskForByte = unchecked((byte)(byte.MaxValue << Math.Min(numberOfEndingBitsToBeOff, bitsPerByte)));
+                var addressByte = addressBytes[addressBytesIndex];
+                if ((addressByte & maskForByte) != addressByte)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static int GetAddressFamilyByteLength(AddressFamily addressFamily)
+            => addressFamily switch
+            {
+                AddressFamily.InterNetwork => IPAddressParserStatics.IPv4AddressBytes,
+                AddressFamily.InterNetworkV6 => IPAddressParserStatics.IPv6AddressBytes,
+                _ => throw new UnreachableException("Unknown address family")
+            };
+        #endregion
+
+        #region Formatting (public)
+        public override string ToString()
+        {
+            return $"{BaseAddress}{addressAndPrefixLengthSeparator}{PrefixLength}";
+        }
+
+        public bool TryFormat(Span<char> destination, out int charsWritten)
+        {
+            if (!BaseAddress.TryFormat(destination, out charsWritten))
+            {
+                charsWritten = 0;
+                return false;
+            }
+
+            if (destination.Length < charsWritten + 2)
+            {
+                return false;
+            }
+
+            destination[charsWritten++] = addressAndPrefixLengthSeparator;
+
+            if (!PrefixLength.TryFormat(destination.Slice(charsWritten), out var prefixLengthCharsWritten))
+            {
+                return false;
+            }
+
+            charsWritten += prefixLengthCharsWritten;
+            return true;
```

Comment:
```suggestion
            return destination.TryWrite($"{BaseAddress}{addressAndPrefixLengthSeparator}{PrefixLength}", out charsWritten);
```
?

Reply:
If we make such a change there wouldn't be a reason to implement ISpanFormattable, since we would allocate a string anyway, which we are trying to avoid

Reply:
This does not allocate a string. That TryWrite method uses a custom string interpolation handler.

Reply:
Oh, didn't know about this method then. Let me look into it and research it more, compare with the current one, thanks!

Reply:
@stephentoub spent some time looking at how TryWrite works and behaves, and seems like we can replace the body of the method with the 1liner you suggested, but probably in future.
The reason is that IPAddress doesn't yet implement ISpanFormattable, even though it has a [TryFormat](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Primitives/src/System/Net/IPAddress.cs#L389) method. And the TryWrite seems to look at whether the type of the object in an interpolation expression implements ISpanFormattable, and if it does - it uses the TryFormat. And since IPAddress doesn't implement ISpanFormattable (yet) - we are ending up using IPAddress.ToString().

So in order to use this one liner we have to have IPAddress implement ISpanFormattable first.

Anyway, doesn't matter how we end up implementing this method - thanks for the suggestion, ended up learning something new about string interpolation

Reply:
I've opened https://github.com/dotnet/runtime/issues/82842.

Reply:
IPAddress now implements ISpanFormattable.

> so we'll need #if to support older runtimes

Yes, though neither IPAddress nor Int32 have a TryFormat in netstandard2.0, nor does ISpanFormattable exist. I suggest if/when we do have such a build that we just not include TryFormat, either.